workmail-save-and-update-email/src/utils.py

```diff
@@ -0,0 +1,237 @@
+                    part.set_charset(html_charset)
+    else:
+        # Its a plain email with text/plain body
+        transfer_encoding = parsed_email['Content-Transfer-Encoding']
+        text_charset = parsed_email.get_content_charset()
+        new_text_body = update_text_content(parsed_email, this_disclaimer_text, this_footer_text)
+        parsed_email.set_content(new_text_body, "plain", charset=text_charset, cte=transfer_encoding)
+    return parsed_email
+
+def download_email(message_id):
+    """
+    This method downloads full email MIME content using GetRawMessageContent API and uses email.parser class
+    for parsing it into Python email.message.EmailMessage class.
+    Reference:
+        https://docs.python.org/3.7/library/email.message.html#email.message.EmailMessage
+        https://docs.python.org/3/library/email.parser.html
+    Parameters
+    ----------
+    message_id: string, required
+        message_id of the email to download
+    Returns
+    -------
+    email.message.Message
+        EmailMessage representation the downloaded email
+    """
+    response = workmail_message_flow.get_raw_message_content(messageId=message_id)
+    email_content = response['messageContent'].read()
+    email_generation_policy = policy.SMTP.clone(refold_source='none')
+    logger.info("Downloaded email from WorkMail successfully")
+    return email.message_from_bytes(email_content, policy=email_generation_policy)
+
+def update_workmail(message_id, content, key):
+    """
+    Uploads the updated message to an S3 bucket in your account and then updates it at WorkMail via
+    PutRawMessageContent API.
+    Reference: https://docs.aws.amazon.com/workmail/latest/adminguide/update-with-lambda.html
+    Parameters
+    ----------
+    message_id: string, required
+        message_id of the email to download
+    content: email.message.Message, required
+         EmailMessage representation the updated email
+    Returns
+    -------
+    None
+    """
+    bucket = os.getenv('UPDATED_EMAIL_BUCKET')
+    if not bucket:
+        raise ValueError("UPDATED_EMAIL_BUCKET not set in environment. Please follow https://docs.aws.amazon.com/lambda/latest/dg/env_variables.html to set it")
+
+    s3.put_object(Body=content.as_bytes(), Bucket=bucket, Key=key)
+    s3_reference = {
+        'bucket': bucket,
+        'key': key
+    }
+    content = {
+        's3Reference': s3_reference
+    }
+    workmail_message_flow.put_raw_message_content(messageId=message_id, content=content)
+    logger.info("Updated email sent to WorkMail successfully")
+
+def save_email(content, key):
+    """
+    Uploads the original message to an S3 bucket in your account
+    """
+    bucket = os.getenv('SAVED_EMAIL_BUCKET')
+    if not bucket:
+        raise ValueError("SAVED_EMAIL_BUCKET not set in environment. Please follow https://docs.aws.amazon.com/lambda/latest/dg/env_variables.html to set it")
+
+    logger.info("saved email bucket: " + bucket)
```

Comment:
It might be useful to log both bucket and key in logs and therefore I suggest to remove this log line and amend the next log line as following

```
logger.info(f"Saved original email at s3://{bucket}/{key} successfully")
```